rust-runtime/aws-smithy-http-server-python/examples/stubgen.py

```diff
@@ -0,0 +1,391 @@
+                if len(parts) != 2:
+                    raise ValueError(
+                        f"Invalid `:rtype` directive: `{line}` must be in `:rtype T:` format"
+                    )
+                rtypes.append(parts[1].rstrip(":"))
+            elif line.startswith(":generic ") and line.endswith(":"):
+                parts = line.split(" ", maxsplit=1)
+                if len(parts) != 2:
+                    raise ValueError(
+                        f"Invalid `:generic` directive: `{line}` must be in `:generic T:` format"
+                    )
+                generics.append(parts[1].rstrip(":"))
+            elif line.startswith(":extends ") and line.endswith(":"):
+                parts = line.split(" ", maxsplit=1)
+                if len(parts) != 2:
+                    raise ValueError(
+                        f"Invalid `:extends` directive: `{line}` must be in `:extends Base[...]:` format"
+                    )
+                extends.append(parts[1].rstrip(":"))
+
+        return DocstringParserResult(
+            types=types,
+            params=params,
+            rtypes=rtypes,
+            generics=generics,
+            extends=extends,
+        )
+
+    @staticmethod
+    def parse_type(obj: Any) -> str:
+        result = DocstringParser.parse(obj)
+        if not result or len(result.types) == 0:
+            return "typing.Any"
+        return result.types[0]
+
+    @staticmethod
+    def parse_function(obj: Any) -> Optional[Tuple[List[Tuple[str, str]], str]]:
+        result = DocstringParser.parse(obj)
+        if not result:
+            return None
+
+        return (
+            result.params,
+            "None" if len(result.rtypes) == 0 else result.rtypes[0],
+        )
+
+    @staticmethod
+    def parse_class(obj: Any) -> Tuple[List[str], List[str]]:
+        result = DocstringParser.parse(obj)
+        if not result:
+            return ([], [])
+        return (result.generics, result.extends)
+
+
+def indent(code: str, level: int) -> str:
+    return textwrap.indent(code, level * " ")
+
+
+def format_doc(obj: Any, indent_level: int) -> str:
+    doc = inspect.getdoc(obj)
+    if not doc:
+        return ""
+
+    # Remove type annotations
+    doc = "\n".join(
+        filter(
+            lambda l: not (
+                (
+                    l.startswith(":type")
+                    or l.startswith(":rtype")
+                    or l.startswith(":param")
+                    or l.startswith(":generic")
+                    or l.startswith(":extends")
+                )
+                and l.endswith(":")
+            ),
+            doc.splitlines(),
+        )
+    ).strip()
+
+    if not doc:
+        return ""
+
+    return indent('"""\n' + doc + '\n"""\n', indent_level)
+
+
+def is_fn_like(obj: Any) -> bool:
+    return (
+        inspect.isbuiltin(obj)
+        or inspect.ismethod(obj)
+        or inspect.isfunction(obj)
+        or inspect.ismethoddescriptor(obj)
+    )
```

Comment:
Don't you need also `inspect.iscoroutinefunction` ?

Reply:
This is what inspect gives you:

```
 'isabstract',
 'isasyncgen',
 'isasyncgenfunction',
 'isawaitable',
 'isbuiltin',
 'isclass',
 'iscode',
 'iscoroutine',
 'iscoroutinefunction',
 'isdatadescriptor',
 'isframe',
 'isfunction',
 'isgenerator',
 'isgeneratorfunction',
 'isgetsetdescriptor',
 'ismemberdescriptor',
 'ismethod',
 'ismethoddescriptor',
 'ismodule',
 'isroutine',
 'istraceback',
```

I am pretty sure you'll have to add `iscoroutine` and `iscoroutinefunction`. I am not sure about `isawaitable`, `isgenerator`, `isgeneratorfunction`, `isroutine`, `isasyncgen` and `isasyncgenfunction`.

This can help you understand what is what:

```python
for x in dir(inspect):
    if x.startswith("is"):
        obj = getattr(inspect, x)
        print(f"============= {x} ============")
        print(f"{obj.__doc__}")
        print("==============================================\n")
```

Reply:
I've added `iscoroutine` and `iscoroutinefunction`. Since `async fn`s are not permitted by PyO3 and coroutines/generators are currently [experimental](https://doc.rust-lang.org/std/ops/trait.Generator.html) we currently only expose regular functions and so the current check should be sufficient.